tensorflow_addons/optimizers/tests/lookahead_test.py

```diff
@@ -186,3 +189,68 @@ def test_serialization():
     config = tf.keras.optimizers.serialize(optimizer)
     new_optimizer = tf.keras.optimizers.deserialize(config)
     assert new_optimizer.get_config() == optimizer.get_config()
+
+
+def _init_model(optimizer, init_w):
+    model = tf.keras.models.Sequential()
+    dense = tf.keras.layers.Dense(input_shape=(3,), units=1)
+    model.add(dense)
+    model.compile(Lookahead(optimizer), loss="mse")
```

Comment:
Also using only `optimizer` here and removing `_optimizer` recursive check your `test_save_load[cpu-no_weights_only-adam` is failing

Reply:
Just to be sure it is not related to https://github.com/tensorflow/tensorflow/issues/40342

Reply:
@WindQAQ What do you think?

Reply:
![image](https://user-images.githubusercontent.com/26186061/91005422-6f064f00-e609-11ea-959e-1d5436e81162.png)
**Fig.** _name of weights at different stages._

For convenience, I only printed the names of the lookahead optimizer weighs. After model weights reloading, "iter" and weights of the inner optimizer are not created yet. In other cases, the weights are complete.

Also please note that lookahead only collects the weighs of its inner optimizer by a property interface, and that is why we can see the weights of the inner optimizer in `lookahead.weights` but they are not checkpointed: https://github.com/tensorflow/addons/blob/f3a50e7e89e23c3b5049af83cbc3697c3c0a3dcb/tensorflow_addons/optimizers/lookahead.py#L135:#L137

Reply:
@qlzh727 Can you expand a little bit over this https://github.com/tensorflow/tensorflow/blame/master/tensorflow/python/keras/engine/training.py#L1986-L1987? Is it just a "preference"?

Reply:
I'am asking this cause it seems to work with this formulation (for "weights_only" case):

```
def test_save_load(optimizer, weights_only):
    x = np.random.standard_normal((10000, 3))
    w = np.random.standard_normal((3, 1))
    y = np.dot(x, w) + np.random.standard_normal((10000, 1)) * 1e-4

    init_w = np.random.standard_normal((3, 1))

    model = _init_model(optimizer, init_w)
    model.fit(x, y, epochs=2, shuffle=False)

    with tempfile.TemporaryDirectory() as ckpt_dir:
        new_model = _init_model(optimizer, init_w)
        new_model.fit(x, y, epochs=1, shuffle=False)
        ckpt_path = os.path.join(ckpt_dir, "ckpt")
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            new_model = None
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            new_model = checkpoint.model

        else:
            new_model.save(ckpt_dir)
            new_model = tf.keras.models.load_model(
                ckpt_dir, custom_objects={"Lookahead": Lookahead}
            )
        new_model.fit(x, y, epochs=1, shuffle=False)
        assert np.allclose(model.predict(x), new_model.predict(x))
```

Reply:
Also in `h5` format is working:

```
def test_save_load(optimizer, weights_only):
    np.random.seed(0)
    x = np.random.standard_normal((10000, 3))
    w = np.random.standard_normal((3, 1))
    y = np.dot(x, w) + np.random.standard_normal((10000, 1)) * 1e-4

    init_w = np.random.standard_normal((3, 1))

    model = _init_model(optimizer, init_w)
    model.fit(x, y, epochs=2, shuffle=False)

    with tempfile.TemporaryDirectory() as ckpt_dir:
        new_model = _init_model(optimizer, init_w)
        new_model.fit(x, y, epochs=1, shuffle=False)
        ckpt_path = os.path.join(ckpt_dir, "ckpt")
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            del new_model
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            new_model = checkpoint.model

        else:
            new_model.save(ckpt_dir + "/test.h5")
            new_model = tf.keras.models.load_model(
                ckpt_dir + "/test.h5"
            )
        new_model.fit(x, y, epochs=1, shuffle=False)
        assert np.allclose(model.predict(x), new_model.predict(x))
```

Reply:
Yet it fails if we also recreate the checkpoint. Your code passes because the old model still stays inside `checkpoint`. `h5` saves model weights by serialization, which is correctly handled, so it could pass.
```
def test_save_load(optimizer, weights_only):
    x = np.random.standard_normal((10000, 3))
    w = np.random.standard_normal((3, 1))
    y = np.dot(x, w) + np.random.standard_normal((10000, 1)) * 1e-4

    init_w = np.random.standard_normal((3, 1))

    model = _init_model(optimizer, init_w)
    model.fit(x, y, epochs=2, shuffle=False)

    with tempfile.TemporaryDirectory() as ckpt_dir:
        new_model = _init_model(optimizer, init_w)
        new_model.fit(x, y, epochs=1, shuffle=False)
        ckpt_path = os.path.join(ckpt_dir, "ckpt")
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            new_model = _init_model(optimizer, init_w)
            new_checkpoint = tf.train.Checkpoint(model=new_model)
            new_checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
        else:
            new_model.save(ckpt_dir)
            new_model = tf.keras.models.load_model(
                ckpt_dir, custom_objects={"Lookahead": Lookahead}
            )
        new_model.fit(x, y, epochs=1, shuffle=False)
        assert np.allclose(model.predict(x), new_model.predict(x))

```

Reply:
I suppose that you need to assign after restore?

```
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            del checkpoint
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            new_model = checkpoint.model
```

Reply:
`checkpoint.model` is `new_model`:
```
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model)
            checkpoint.save(file_prefix=ckpt_path)
            new_model = _init_model(optimizer, init_w)
            new_checkpoint = tf.train.Checkpoint(model=new_model)
            new_checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
            assert new_model is new_checkpoint.model
```

Reply:
Ok so it seems that the inner optimizer is not tracked automatically also with explicit ceckpoint API:

```
        if weights_only:
            checkpoint = tf.train.Checkpoint(model=new_model, optimizer=new_model.optimizer._optimizer)
            checkpoint.save(file_prefix=ckpt_path)
            del checkpoint
            del new_model
            new_model = _init_model(optimizer, init_w)
            checkpoint = tf.train.Checkpoint(model=new_model,optimizer=new_model.optimizer._optimizer)
            checkpoint.restore(tf.train.latest_checkpoint(ckpt_dir))
```

Reply:
@BinyanHu Can you start to check your `weights_only` case with https://github.com/tensorflow/addons/pull/2126?

Reply:
`weights_only=True` passed on that branch, yet `weights_only=False` still fails.

Reply:
@BinyanHu Yes for the `SavedModel` case I suspect that this is still vaild https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/saving/saved_model/load.py#L99-L104 (especially on optimizers with slots).

Reply:
Sure. So after the tracking of inner optimizer weights is fixed, the remaining parts should be tensorflow issues. Besides, similar wrapper optimizers such as `MovingAverage` should also be fixed in this way. Anything left to do with this pull request?

Reply:
Yes `AvarageWrapper` was added.